### PR TITLE
Fix reentrant hang issue

### PR DESCRIPTION
**Pull request by simonjiao**
* **make sure that you have executed all the following process and no errors occur**
  - [ ] make fmt
  - [ ] make lint
  - [ ] make test

* **The major changes of this PR**
Fix two following issues:
1. Transaction receipt is missing when EVM returns error code 1
2. Nonce is not increased on revert.

* **The major impacts of this PR**
  - [ ] Impact WASM?
  - [ ] Impact Web3 API?
  - [ ] Impact mainnet data compatibility?

* **Extra documentations**

### CODE REVIEW

File: tools/devnet/startnodes.sh [hunk 1]
```diff
@@ -26,7 +21,7 @@ if [ -z "$Node" ] || ([ ! -z "$Node" ] && [ "$Node" = "$node" ]); then
         ENABLE_LEDGER_SERVICE=true \
         ENABLE_ETH_API_SERVICE=true \
         ENABLE_QUERY_SERVICE=true \
-        $BIN/abcid $DEVNET/$node >> $DEVNET/$node/abcid.log 2>&1  &
+        abcid $DEVNET/$node >> $DEVNET/$node/abcid.log 2>&1  &
```
Comment (ws4charlie) [inline tools/devnet/startnodes.sh:24]: This $BIN is defined in env.sh and can helps switching between `debug` and `release` versions.

File: src/components/contracts/baseapp/src/extensions.rs [hunk 2]
```diff
@@ -48,6 +50,46 @@ impl SignedExtension for CheckNonce {
...
+        if ctx.header.height >= CFG.checkpoint.evm_checktx_nonce && result.code != 0 {
+            let prev = result
+                .source
+                .as_ref()
+                .map(|who| module_account::App::<BaseApp>::nonce(ctx, who))
+                .unwrap_or_default();
+
+            ctx.state.write().discard_session();
+
+            let current = result
```
Comment (ws4charlie) [inline src/components/contracts/baseapp/src/extensions.rs:68]: Can we add some comments: If EVM transaction fails, state change gets reverted/discarded except nonce.

**Review by ws4charlie — APPROVED** [review]

File: src/components/contracts/primitives/core/src/transaction.rs [hunk 3]
```diff
@@ -167,12 +173,16 @@ where
                     res.log = String::from("ctx state is not good to commit");
 
                     ctx.state.write().discard_session();
-                } else if res.code == 0 {
-                    Extra::post_execute(ctx, pre, &res)?;
-
-                    ctx.state.write().commit_session();
+                } else if signed_tx {
+                    if res.code == 0 {
+                        Extra::post_execute(ctx, pre, &res)?;
```
Comment (ws4charlie) [inline src/components/contracts/primitives/core/src/transaction.rs:178]: Just wanted to confirm if below 2 statement are correct or not
1> Extra::post_execute does nothing
2> U::post_execute does the nonce correction

**Review by ws4charlie — APPROVED** [review]
looks good

File: src/components/contracts/primitives/core/src/transaction.rs (hunk 3 above)
Comment (simonjiao) [reply src/components/contracts/primitives/core/src/transaction.rs:178]: Yes. For Transactions which implement Executable trait, only ethereum txns could perform inc_nonce. And for a signed txn, ethereum module will do nothing but fail the txn immediately.

**Review by ws4charlie — APPROVED** [review]

**Pull request merged**